src/FileType_XDMF.f90

```diff
@@ -0,0 +1,1236 @@
+    ENDDO
+
+    ! If some children are grids, it is not a leaf
+    IF(grid_ctr > 0) THEN
+      ! Allocate children for current mesh and create entities
+      ALLOCATE(mesh%children(grid_ctr))
+      mesh_ctr=1
+      ! Recursively create children of each grid
+      DO i=1,SIZE(xmle_children)
+        IF(xmle_children(i)%name%upper() == 'GRID') THEN
+          strIn='Name'
+          CALL xmle_children(i)%getAttributeValue(strIn,strOut)
+          mesh%children(mesh_ctr)%name = strOut
+          mesh%children(mesh_ctr)%parent => mesh
+          CALL create_XDMFMesh_from_file(mesh%children(mesh_ctr), xmle_children(i), h5)
+          mesh_ctr = mesh_ctr + 1
+        ENDIF
+      ENDDO
+    ! If this mesh does not have grid children it is a leaf on the tree.
+    ! Add vertices, cells, etc.
+    ELSE
+      CALL setup_leaf_XDMFMesh_from_file(mesh, xmle, h5)
+    ENDIF
+  ELSE
+    CALL eXDMF%raiseError(modName//'::'//myName// &
+      ' - Expected the XML element to have children.')
+  ENDIF
+ENDSUBROUTINE create_XDMFMesh_from_file
+!
+!-------------------------------------------------------------------------------
+!> @brief Setup the leaf mesh objects which contain vertices, cells, etc.
+!> @param mesh the parent mesh
+!> @param xmle the child XML element
+!> @param h5 the HDF5 file containing mesh data
+!>
+SUBROUTINE setup_leaf_XDMFMesh_from_file(mesh, xmle, h5)
+  CHARACTER(LEN=*),PARAMETER :: myName='setup_leaf_XDMFMesh_from_file'
+  TYPE(XDMFMeshType),INTENT(INOUT),TARGET  :: mesh
+  TYPE(XMLElementType), INTENT(INOUT) :: xmle
+  TYPE(HDF5FileType), INTENT(INOUT) :: h5
+  TYPE(XMLElementType), POINTER :: xmle_children(:), ele_children(:)
+  TYPE(StringType) :: elname, strIn, strOut, content, group, dtype, toponame, &
+    xdmf_id_str
+  TYPE(StringType),ALLOCATABLE :: strArray(:),segments(:)
+  INTEGER(SLK) :: nverts, ncells, xdmf_id,vert_ctr,i,j
+  INTEGER(SNK) :: ncell_sets
+  INTEGER(SNK),ALLOCATABLE :: dshape(:)
+  REAL(SSK),ALLOCATABLE :: vals4_2d(:,:)
+  REAl(SDK),ALLOCATABLE :: vals8_2d(:,:)
+  INTEGER(SNK),ALLOCATABLE :: ivals4_1d(:),ivals4_2d(:,:)
+  INTEGER(SLK),ALLOCATABLE :: ivals8_1d(:),ivals8_2d(:,:)
+  TYPE(XDMFCellSet), ALLOCATABLE :: cell_sets_temp(:)
+
+  REQUIRE(xmle%hasChildren())
+  CALL xmle%getChildren(xmle_children)
+  ! Each XML element has a type of information.
+  ! Handle each with a CASE
+  DO i=1,SIZE(xmle_children)
+    elname=xmle_children(i)%name%upper()
+    SELECTCASE(ADJUSTL(elname))
+      CASE("GEOMETRY")
+        ! GeometryType
+        strIn='GeometryType'
+        CALL xmle_children(i)%getAttributeValue(strIn,strOut)
+        IF(strOut /= 'XYZ') THEN
+          CALL eXDMF%raiseWarning(modName//'::'//myName// &
+            ' - GeometryType only supports XYZ right now.')
+        ENDIF
+        ! Format
+        CALL xmle_children(i)%getChildren(ele_children)
+        REQUIRE(SIZE(ele_children) == 1)
+        strIn='Format'
+        CALL ele_children(1)%getAttributeValue(strIn,strOut)
```

Comment:
Instead of indexing children(1) throughout, add a named INTEGER. Is (1) the parent or a top level? Could do PARENT_IDX.

Reply:
The (1) refers to the only child in a lot of places. I feel (1) is better than (ONLY_CHILD) or something like that.